[quant] Implement APoT fake quantization

asl3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79845
* #80364

### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

asl3 committed:
[quant] Implement APoT fake quantization

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

dzdang commented on test/quantization/core/experimental/test_fake_quantize.py:
```diff
@@ -0,0 +1,58 @@
+        max_val = torch.max(X)
+
+        observer = APoTObserver(b=4, k=2)
+        qparams = observer.calculate_qparams(signed=False, min_val=min_val, max_val=max_val)
```
unpack here directly for clarity? e.g.,
gamma, alpha, aquantized_levels.... = .....

dzdang commented on test/quantization/core/experimental/test_fake_quantize.py:
```diff
@@ -0,0 +1,58 @@
+        X = 1000 * torch.rand(20)
+
+        min_val = torch.min(X)
+        max_val = torch.max(X)
```
we can just use https://pytorch.org/docs/stable/generated/torch.aminmax.html#torch.aminmax

dzdang commented on torch/ao/quantization/experimental/fake_quantize.py:
```diff
@@ -0,0 +1,29 @@
+    def forward(self, X: torch.Tensor, signed: bool):
+        if self.observer_enabled[0] == 1:
+            min_val = torch.min(X)
+            max_val = torch.max(X)
```
aminmax

dzdang reviewed (commented):
had a few minor suggestions

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


`Summary:`
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

`Test Plan:`
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/experimental/test_fake_quantize.py:
```diff
@@ -0,0 +1,85 @@
+        min_val, max_val = torch.aminmax(X)
+
+        observer = APoTObserver(b=4, k=2)
+        alpha, gamma, quantization_levels, level_indices = observer.calculate_qparams(signed=False,
```
do we need to reset `observer` or create a new observer with the same parameters here?

jerryzh168 replied:
I still feel we should not expose the api for user to set `min_val`/`max_val` since it's just for testing, if you actually need this feature, we can just manually set `min_val` and `max_val`, like:
observer.min_val = xxx and observer.max_val = xxx.

Exposing this to calculate_qprams means that calculate_qparams not only calculate qparams but also may change the state of observer (min_val/max_val), which is not desirable

asl3 replied:
agreed, updated `calculate_qparams` method signature in separate PR

jerryzh168 commented on test/quantization/core/experimental/test_fake_quantize.py:
```diff
@@ -0,0 +1,85 @@
+
+        X_reduced_precision_fp = apot_fake.forward(torch.clone(X), False)
+
+        X_expected = X.apply_(lambda x: float_to_reduced_precision(x, quantization_levels, level_indices))
```
float_to_reduce_precision is the same as "quantize + dequantize"? maybe we can rename that to make it clearer?

asl3 replied:
yes sounds good, will do!

jerryzh168 commented on torch/ao/quantization/experimental/fake_quantize.py:
```diff
@@ -0,0 +1,32 @@
+
+        if self.observer_enabled[0] == 1:
+            self.observer.forward(X)
+            alpha, gamma, quantization_levels, level_indices = self.observer.calculate_qparams(signed)
```
if we guard this with `self.observer_enabled[0] == 1`, we should also store quantization paramters in the fake quantize module I think: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fake_quantize.py#L114-L115, otherwise when we do not enable observer, fake quantize would also fail since these values are initialized to None

jerryzh168 commented on torch/ao/quantization/experimental/fake_quantize.py:
```diff
@@ -0,0 +1,32 @@
+                    and quantization_levels is not None
+                    and level_indices is not None), "Must set qparams for fake quant"
+
+            quantizer = APoTQuantizer(alpha, gamma, quantization_levels, level_indices)
```
quantizer is not used?

jerryzh168 commented on torch/ao/quantization/experimental/fake_quantize.py:
```diff
@@ -0,0 +1,32 @@
+
+            quantizer = APoTQuantizer(alpha, gamma, quantization_levels, level_indices)
+            X = quantize_APoT(X, alpha, gamma, quantization_levels, level_indices)
+            X = dequantize_APoT(X, alpha, gamma, quantization_levels, level_indices)
```
to align with C++ api I think dequantize_APoT should not take quantization parameters since they are stored in X.quantizer already

jerryzh168 reviewed (changes requested):
looks good, please see some comments inline

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

dzdang commented on torch/ao/quantization/experimental/fake_quantize.py:
```diff
@@ -0,0 +1,35 @@
+
+    def __init__(self, observer: APoTObserver):
+        super().__init__()
+        self.activation_post_process = observer
```
need constructor args?

jerryzh168 replied:
yeah, original code is probably a bit confusing, maybe we can rename `observer` to `observer_cls` or `observer_constructor` to make it clearer, we do need to instantiate the observer in `__init__.py`: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fake_quantize.py#L133

dzdang commented on torch/ao/quantization/experimental/fake_quantize.py:
```diff
@@ -0,0 +1,35 @@
+    def __init__(self, observer: APoTObserver):
+        super().__init__()
+        self.activation_post_process = observer
+        self.alpha = self.gamma = self.quantization_levels = self.level_indices = Tensor()
```
why is this initialized to empty tensor instead of attriubtes in observer? this makes the forward method faulty

asl3 replied:
@dzdang i initialize these qparams in forward by calling `calculate_qparams` , like https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fake_quantize.py#L158 (here the qparams are not set in init). should i be calling `calculate_qparams` in my constructor instead?

dzdang replied:
what if `if self.observer_enabled[0] == 1`? does your workflow guaranteed the qparams are set before you come across the case self.observer_enabled[0] == 1?

asl3 replied:
@dzdang In the case `self.fake_quant_enabled[0] == 1` , I have an assert statement to check that the qparams have been set before I use the qparams.
I wasn't sure how to initialize elsewhere besides in the `self.observer_enabled[0] == 1` case in `forward` since if observer is not enabled shouldn't I not be calling `observer.calculate_qparam` to set the qparams? I thought this made most sense and was most consistent with the uniform fake_quantize but lmk what you think

dzdang replied:
this is related to the constructor args comment. if you create an observer in the constructor, then couldn't you just assign the fake parameter's qparams to the observer's?

asl3 replied:
do you mean assigning qparams what is returned by `observer.calculate_qparams` ? observer does not store qparams, just `b`, `k`, `n`, `max_val`, `min_val`. calculate_qparams also takes an additional arg, `signed`

dzdang replied:
oh right. hmm I think what you've done is fine then.

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

dzdang commented on test/quantization/core/experimental/test_fake_quantize.py:
```diff
@@ -0,0 +1,76 @@
+
+        # get X_expected by converting fp -> apot -> fp to simulate quantize -> dequantize
+        X_expected = X.apply_(lambda x: float_to_apot(x, quantization_levels, level_indices))
+        X_expected = X_expected.apply_(lambda x: apot_to_float(x, quantization_levels, level_indices))
```
can we call `quantize_APoT` and `dequantize_APoT` instead of using lambdas?

dzdang reviewed (approved):
had one comment, but i think looks good otherwise

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement APoT fake quantization"


### Summary:
This PR implements APoT fake quantization for the purpose of quantization aware training. This implements `calculate_qparams` and `forward `methods to be used in fake quantization.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g
